src/compiler/scala/tools/nsc/transform/patmat/Logic.scala

```diff
@@ -174,8 +174,8 @@ trait Logic extends Debugging  {
     def simplify(f: Prop): Prop = {
 
       // limit size to avoid blow up
-      def hasImpureAtom(ops: Seq[Prop]): Boolean = ops.size < 10 &&
-        ops.combinations(2).exists {
+      def hasImpureAtom(ops: Traversable[Prop]): Boolean =
+        ops.toSeq.combinations(2).exists {
```

Comment:
this looks tunable
It looks to be O n squared, and will generate lots of local garbage

could you scan for all Not(_) and if you find any scan for matching content?
Or something similar to avoid all of the list generation and unapplySeq

Reply:
The garbage problem could be fixed with a custom iterator:

```
class PairsIterator[A](as: Seq[A]) {
  private val asArray = as.toArray
  var indexA, indexB: Int = 0
  def hasNext: Boolean = ...
  def advance(): Unit = ...
  def _1: A = asArray(indexA)
  def _2: B = asArray(indexA)
}
```